projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,256 @@
+                LOGD << " we have a valid measurement";
+
+            } else {
+                numTrials--;
+            }
+
+        } catch(BoBRobotics::GPS::GPSError &e) {
+            LOGW << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]";
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+
+    if (numTrials == 0) {
+        LOGW << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program ";
+        return EXIT_FAILURE;
+    }
+    std::stringstream ss;
+    ss << "imgdataset_"  << str << "_" << gps_hour << "-" << gps_minute << "-" << gps_second;
+    const auto folderName = ss.str();
+    const char *c  = folderName.c_str();
+    BOB_ASSERT(mkdir(c, 0777) == 0);
+    LOGD << "directory " << c << " created";
+    auto cam = getPanoramicCamera();
+    const auto cameraRes = cam->getOutputSize();
+    LOGD << "camera initialised " << cameraRes;
+
+
+
+    //--------------------------------->>>>>>>>>>> START RECORDING <<<<<<<<<<<<----------------------------
+    // start camera thread
+    cv::Mat tmpImage;
+    std::thread readCameraThread(&readCameraThreadFunc, cam.get(), std::ref(tmpImage), std::ref(shouldQuit), std::ref(mex));
+    std::this_thread::sleep_for(std::chrono::milliseconds(4000));
+    BoBRobotics::BackgroundExceptionCatcher catcher;
+    catcher.trapSignals();
+
+    int i = 0;
+    BoBRobotics::Stopwatch sw, sw_timestamp;
+    std::array<degree_t, 3> angles;
+    // for x timestep - starting stopwatch
+    sw_timestamp.start();
+
+    std::ofstream coordinates; // file handle
+    std::ostringstream folderString;
+    folderString << folderName << "/coordinates.csv";
+    //  headers:   speed|steering_angle|gpsQuality|latitude|longitude|altitude|roll|pitch|yaw|imagename|timestamp
+    //write headers  for csv
+    coordinates.open (folderString.str(), std::ofstream::trunc);
+    coordinates << "speed" << "," << "steering angle" << "," << "gps quality" << "," << "lat" << "," << "lon" << "," << "altitude"  << "," << "roll" << "," << "pitch" << "," << "yaw"  << ",";
+    coordinates << "image name" << "," << "timestamp" <<"\n";
+
+    for (;;) {
+        // poll from camera thread
+        sw.start();
+        cv::Mat originalImage;
+        mex.lock();
+        tmpImage.copyTo(originalImage);
+        mex.unlock();
+        std::ostringstream fileString;
+        fileString << folderName << "/image" << i << ".jpg";
+        std::string fileName = fileString.str();
+
+
+        if (!cv::imwrite(fileName, originalImage)) {
+            LOGW << "Could not save image file";
+        }
+        // get imu data
+        float roll,pitch,yaw;
+        try {
+            angles = imu.getEulerAngles();
+            yaw = angles[0].value();
+            pitch = angles[1].value();
+            roll = angles[2].value();
+        } catch(...) {
```

Comment:
I'd do:
```c++
} catch (std::exception &e) {
    LOGE << "Could not read from IMU: " << e.what();
   // ...
```

The problem with not printing anything is that afterwards you can't distinguish the case where there's a bug in ``getEulerAngles()`` from the case where the hardware is playing up. I know it'll probably always be the latter, but it could be frustrating if we have a CSV file of NaNs and can't diagnose what's gone wrong.